### PR TITLE
optimistic: fix optimist panic after table/source removed

### LINKED ISSUE #1970: restart master will panic after some table remove from optimistic task
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
```
start a task with 2 sources
source1: alter table tb1 add column new_col1 int
stop-task
update task configuration, add tb1 to block-list
star-task
restart master
```
2. What did you expect to see?
Task is running normally
3. What did you see instead?
master panic
```
goroutine 349 [running]:
github.com/pingcap/dm/pkg/shardddl/optimism.(*LockKeeper).RebuildLocksAndTables(0xc00055e1c0, 0xc000278000, 0xc000ab7590, 0xc0007acae0, 0xc001800630, 0xc001800540, 0xc001800660)
        /nvme1n1/gmhdbjd/code/dm/pkg/shardddl/optimism/keeper.go:63 +0x8d9
github.com/pingcap/dm/dm/master/shardddl.(*Optimist).recoverLocks(0xc00053c140, 0xc000ab7590, 0xc000b80b10, 0xc0007acae0, 0xc000b80d80, 0x0, 0x0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/shardddl/optimist.go:387 +0x3aa
github.com/pingcap/dm/dm/master/shardddl.(*Optimist).rebuildLocks(0xc00053c140, 0x3c, 0x3c, 0x3c, 0x0, 0x0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/shardddl/optimist.go:267 +0x1212
github.com/pingcap/dm/dm/master/shardddl.(*Optimist).Start(0xc00053c140, 0x52bfcb0, 0xc000551ac0, 0xc000278000, 0x0, 0x0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/shardddl/optimist.go:75 +0x21f
github.com/pingcap/dm/dm/master.(*Server).startLeaderComponent(0xc000228540, 0x52bfcb0, 0xc000551ac0, 0x0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/election.go:185 +0x809
github.com/pingcap/dm/dm/master.(*Server).electionNotify(0xc000228540, 0x52bfcb0, 0xc000551ac0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/election.go:71 +0xd1b
github.com/pingcap/dm/dm/master.(*Server).Start.func3(0xc000228540, 0x52bfcb0, 0xc000551ac0)
        /nvme1n1/gmhdbjd/code/dm/dm/master/server.go:225 +0x10b
created by github.com/pingcap/dm/dm/master.(*Server).Start
        /nvme1n1/gmhdbjd/code/dm/dm/master/server.go:223 +0x1de8
```
4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        master
        ```

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
close #1970 

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

**Comment by GMHDBJD** [comment]
/run-all-tests
```execute statement failed: ALTER TABLE `shardddl`.`tb` ADD COLUMN `new_col1` INT, RawCause: Error 8028: Information schema is changed during the execution of the statement(for example, table definition may be updated by other DDL ran in parallel)```

File: pkg/shardddl/optimism/keeper.go [hunk 1]
```diff
@@ -51,15 +51,23 @@ func (lk *LockKeeper) RebuildLocksAndTables(
...
+			for schema, schemaInfos := range sourceInfos {
+				for table, info := range schemaInfos {
 					lockID := utils.GenDDLLockID(info.Task, info.DownSchema, info.DownTable)
 					if lock, ok = lk.locks[lockID]; !ok {
-						lk.locks[lockID] = NewLock(cli, lockID, info.Task, info.DownSchema, info.DownTable, lockJoined[lockID], lockTTS[lockID])
-						lock = lk.locks[lockID]
+						lock = NewLock(cli, lockID, info.Task, info.DownSchema, info.DownTable, lockJoined[lockID], lockTTS[lockID])
 					}
+					// filter info which doesn't have SourceTable
+					// SourceTable will be changed after user update block-allow-list
```
Comment (Ehco1996) [inline pkg/shardddl/optimism/keeper.go:63]: as opposed to cleaning up error info here
maybe we can celar outated info in etcd after user update `block-allow-list` is a better idea?

**Review by lance6716 — APPROVED** [review]

File: pkg/shardddl/optimism/keeper.go (hunk 1 above)
Comment (GMHDBJD) [reply pkg/shardddl/optimism/keeper.go:63]: Not reliable in current design.
```
user do a ddl, 
stop-task
drop table
redo task(start-task remove meta)
```
the ddl info still in etcd, but sourceTable was deleted.

Comment (Ehco1996) [reply pkg/shardddl/optimism/keeper.go:63]: can we clear data when `remove-meta`?

Comment (GMHDBJD) [reply pkg/shardddl/optimism/keeper.go:63]: It's a idea

**Pull request merged**